openedx_learning/apps/authoring/collections/api.py

```diff
@@ -204,3 +205,37 @@ def get_collections(learning_package_id: int, enabled: bool | None = True) -> Qu
     if enabled is not None:
         qs = qs.filter(enabled=enabled)
     return qs.select_related("learning_package").order_by('pk')
+
+
+def remove_unpublished_from_collections(learning_package_id: int) -> None:
+    """
+    Removes all unpublished entities from collections for a
+    given learning package.
+
+    One use case for this function is when reverting to published.
+    If a collection has unpublished entities and the learning package is reverted
+    to the published version, the unpublished entities need to be removed.
```

Comment:
I think this API call would have unintended side-effects, since there may be many Collections that have unpublished things in them. Doing a revert shouldn't empty them all out.

Reply:
> Doing a revert shouldn't empty them all out.

I understand that, in the context of libraries, it should be empty if no component is published. When a library is published, all its components, including those belonging to collections, are published. Similarly, when reverting, all components, including those belonging to collections, are discarded.

Since it is in the context of libraries, this function is only called when [reverting a library](https://github.com/openedx/edx-platform/pull/35734/files#diff-a760c3e87f2f671ac2453abbbaf6f10ca052c7acf1c46c1d36f24646e1d52a00R1333), and has not been added when [reverting any learning_package](https://github.com/openedx/openedx-learning/blob/46140b1e28c227173f3098d6d5d934f13503e63a/openedx_learning/apps/authoring/publishing/api.py#L422)

I updated the comment to specify that the use case is in the context of libraries: https://github.com/openedx/openedx-learning/pull/252/commits/e16f6a2671bbfeb0d81facd49b1b438f4830a37f

Reply:
If a library has components that were never published, but they've been added to a collection: when you "revert changes" to the library, those components should be essentially deleted. In that case, the deletion should already be propagating to the collection model. If not, that's a bug we need to fix in here - but shouldn't require an explicit API call.

If a library has components that were published then modified and/or added to new collections: when you "revert changes" to the library, the changes to the components should be reset so they match the published version. But, they will still be in any new collections that they've been added to. This is because tags and collections are not part of the draft-publish workflow.

Does that make sense? If so, I don't think we need this proposed API.

Reply:
@bradenmacdonald: I think we worked this out in the issue thread: https://github.com/openedx/modular-learning/issues/234#issuecomment-2455417567